File: 8-context-managers/scope_timer.cpp
#include <chrono>
#include <iostream>
#include <string_view>


class ScopeTimer {
public:
    using Clock = std::chrono::steady_clock;
    static_assert(Clock::is_steady);

    std::string_view name;
    Clock::time_point start_time;
    Clock::time_point end_time;

    ScopeTimer(std::string_view name) :
        name{name},
        start_time{Clock::now()},
        end_time{}
    {}

    ~ScopeTimer() {
        end_time = Clock::now();
        _process_time();
    }

private:
    void _process_time() const {
        auto const elapsed_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(end_time - start_time);
        std::cout << "ScopeTimer \"" << name << "\" took " << elapsed_time_ns.count() << "ns\n";
    }
};


void database_query() {
    // ...
    {
        ScopeTimer const timer{"database query"};
        // Send query to database
    }
    // ...
}


int main() {
    database_query();
}

File: 8-context-managers/scope_timer.py
import time


class ScopeTimer:
    def __init__(self, name: str) -> None:
        self.name = name
        self.start_time: float | None = None
        self.end_time: float | None = None

    def __enter__(self) -> 'ScopeTimer':
        self.start_time = time.perf_counter_ns()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.end_time = time.perf_counter_ns()
        self._process_time()
    
    def _process_time(self) -> None:
        assert self.start_time is not None
        assert self.end_time is not None
        elapsed_time_ns = self.end_time - self.start_time
        print(f'ScopeTimer "{self.name}" took {elapsed_time_ns}ns')


def database_query():
    ...
    with ScopeTimer('database query'):
        # Send query to database
        ...
    ...


database_query()
